lib/triggers/streamPlatformEventsPubSub.js

```diff
@@ -0,0 +1,157 @@
+/* eslint-disable no-use-before-define, no-unused-vars */
+const { messages } = require('elasticio-node');
+const commons = require('@elastic.io/component-commons-library');
+const { callJSForceMethod } = require('../helpers/wrapper');
+const { getSecret, refreshToken } = require('../util');
+const { PubSubClient } = require('../PubSubClient');
+
+let client;
+let context;
+let refreshTokenNeeded = false;
+let authFailureRetry = 1;
+let connectFailureRetry = 1;
+let nextStartReplayId;
+const MAX_AUTH_RETRIES = 2;
+const MAX_CONNECT_RETRIES = 5;
+let eventEmitter;
+let creationInProgress = false;
+
+const reconnectHandler = async (reconnectSleepTime, cfg) => {
+  if (client) client.close();
+  creationInProgress = false;
+  client = undefined;
+  await commons.sleep(reconnectSleepTime);
+  await processTrigger.call(context, {}, cfg, { nextStartReplayId });
+};
+
+let healthCheckRunning = false;
+const healthCheck = (cfg) => {
+  if (client.getLastKeepAlive() < (new Date().getTime() - 300000)) {
+    context.logger.warn('Last KeepAlive was more than 5 minutes ago, going to restart trigger');
+    healthCheckRunning = false;
+    reconnectHandler(20000, cfg);
+  } else {
+    if (!healthCheckRunning) setTimeout(() => healthCheck(cfg), 1000);
+    healthCheckRunning = true;
+  }
+};
+
+const isNumberNaN = (num) => Number(num).toString() === 'NaN';
+
+async function processTrigger(msg, cfg, snapshot) {
+  context = this;
+  if (client) {
+    this.logger.info(`Subscription is still running, last keepalive from Salesforce ${new Date(client.getLastKeepAlive()).toISOString()}, waiting for new messages`);
+    client.setLogger(context.logger);
+    return;
+  }
+  if (creationInProgress) {
+    this.logger.info('Subscription recreate in progress');
+    return;
+  }
+  creationInProgress = true;
+  const { secretId } = cfg;
+  let { eventCountPerRequest = 10, initialReplayId } = cfg;
+
+  if (eventCountPerRequest && (isNumberNaN(eventCountPerRequest) || Number(eventCountPerRequest) <= 0 || Number(eventCountPerRequest) > 100)) {
+    throw new Error('"Number of events per request" must be valid number between 1 and 100');
+  }
+  if (initialReplayId && (isNumberNaN(eventCountPerRequest) || Number(eventCountPerRequest) <= 0)) {
+    throw new Error('"Start from Replay Id" must be a valid number above 0');
+  }
+
+  if (eventCountPerRequest) eventCountPerRequest = Number(eventCountPerRequest);
+  if (initialReplayId) initialReplayId = Number(initialReplayId);
+
+  if (!secretId) {
+    context.logger.error('secretId is missing in configuration, credentials cannot be fetched');
+    creationInProgress = false;
+    throw new Error('secretId is missing in configuration, credentials cannot be fetched');
+  }
+  if (refreshTokenNeeded) {
+    await refreshToken(context, secretId, msg.id);
+    refreshTokenNeeded = false;
+  }
+  context.logger.info('Starting "Subscribe to PubSub" trigger');
```

Comment:
This should come at the very beginning of the trigger

Reply:
moved